File: project/.sdk/tm/lua/utility/fetcher.lua
-- ProjectName SDK utility: fetcher

local vs = require("utility.struct.struct")
local json = require("dkjson")

local function default_http_fetch(fullurl, fetchdef)
  -- Use LuaSocket or similar for real HTTP requests.
  -- This is the default live implementation.
  local http_ok, http = pcall(require, "socket.http")
  local ltn12_ok, ltn12 = pcall(require, "ltn12")

  if not http_ok or not ltn12_ok then
    return nil, "HTTP library not available (install luasocket)"
  end

  local method = fetchdef["method"] or "GET"
  local body_str = fetchdef["body"]
  local headers = fetchdef["headers"] or {}

  if type(body_str) ~= "string" then
    body_str = nil
  end

  if body_str ~= nil then
    headers["content-length"] = tostring(#body_str)
  end

  local response_body = {}
  local res, code, response_headers = http.request({
    url = fullurl,
    method = method,
    headers = headers,
    source = body_str and ltn12.source.string(body_str) or nil,
    sink = ltn12.sink.table(response_body),
  })

  if not res then
    return nil, code
  end

  local body = table.concat(response_body)
  local resp_headers = {}
  if type(response_headers) == "table" then
    for k, v in pairs(response_headers) do
      resp_headers[string.lower(k)] = v
    end
  end

  local json_body = nil
  if #body > 0 then
    json_body = json.decode(body)
  end

  local status_text = "OK"
  if code >= 400 then
    status_text = "Error"
  end

  return {
    status = code,
    statusText = status_text,
    headers = resp_headers,
    json = function() return json_body end,
    body = body,
  }, nil
end


local function fetcher_util(ctx, fullurl, fetchdef)
  if ctx.client.mode ~= "live" then
    return nil, ctx:make_error("fetch_mode_block",
      'Request blocked by mode: "' .. ctx.client.mode ..
      '" (URL was: "' .. fullurl .. '")')
  end

  local options = ctx.client:options_map()
  if vs.getpath(options, "feature.test.active") == true then
    return nil, ctx:make_error("fetch_test_block",
      'Request blocked as test feature is active' ..
      ' (URL was: "' .. fullurl .. '")')
  end

  local sys_fetch = vs.getpath(options, "system.fetch")

  if sys_fetch == nil then
    return default_http_fetch(fullurl, fetchdef)
  end

  if type(sys_fetch) == "function" then
    return sys_fetch(fullurl, fetchdef)
  end

  return nil, ctx:make_error("fetch_invalid", "system.fetch is not a valid function")
end

return fetcher_util

File: project/.sdk/tm/py/utility/fetcher.py
# ProjectName SDK utility: fetcher

from __future__ import annotations
import json
from utility.voxgig_struct import voxgig_struct as vs


# Default User-Agent — many CDNs (notably Cloudflare) reject requests with
# Python's default urllib UA ("Python-urllib/3.x"), returning 403 before
# the request even reaches the origin. Set a Mozilla-shaped UA so the SDK
# behaves like every other HTTP client by default. Users can still override
# by passing a User-Agent header in fetchdef.
_DEFAULT_USER_AGENT = "Mozilla/5.0 (compatible; ProjectNameSDK/1.0)"


def _default_http_fetch(fullurl, fetchdef):
    import urllib.request
    import urllib.error

    method = fetchdef.get("method", "GET")
    body_str = fetchdef.get("body")
    headers = fetchdef.get("headers", {})

    if not isinstance(body_str, str):
        body_str = None

    data = body_str.encode("utf-8") if body_str is not None else None

    req = urllib.request.Request(fullurl, data=data, method=method)
    has_ua = False
    for k, v in headers.items():
        if k.lower() == "user-agent":
            has_ua = True
        req.add_header(k, v)
    if not has_ua:
        req.add_header("User-Agent", _DEFAULT_USER_AGENT)

    try:
        resp = urllib.request.urlopen(req)
        body = resp.read().decode("utf-8")
        resp_headers = {}
        for k, v in resp.getheaders():
            resp_headers[k.lower()] = v

        json_body = None
        if len(body) > 0:
            try:
                json_body = json.loads(body)
            except Exception:
                pass

        status = resp.getcode()
        status_text = "OK" if status < 400 else "Error"

        return {
            "status": status,
            "statusText": status_text,
            "headers": resp_headers,
            "json": lambda: json_body,
            "body": body,
        }, None
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8")
        resp_headers = {}
        for k, v in e.headers.items():
            resp_headers[k.lower()] = v

        json_body = None
        if len(body) > 0:
            try:
                json_body = json.loads(body)
            except Exception:
                pass

        status_text = "OK" if e.code < 400 else "Error"

        return {
            "status": e.code,
            "statusText": status_text,
            "headers": resp_headers,
            "json": lambda: json_body,
            "body": body,
        }, None
    except Exception as e:
        return None, str(e)


def fetcher_util(ctx, fullurl, fetchdef):
    if ctx.client.mode != "live":
        return None, ctx.make_error("fetch_mode_block",
            'Request blocked by mode: "' + ctx.client.mode +
            '" (URL was: "' + fullurl + '")')

    options = ctx.client.options_map()
    if vs.getpath(options, "feature.test.active") is True:
        return None, ctx.make_error("fetch_test_block",
            'Request blocked as test feature is active'
            ' (URL was: "' + fullurl + '")')

    sys_fetch = vs.getpath(options, "system.fetch")

    if sys_fetch is None:
        return _default_http_fetch(fullurl, fetchdef)

    if callable(sys_fetch):
        return sys_fetch(fullurl, fetchdef)

    return None, ctx.make_error("fetch_invalid", "system.fetch is not a valid function")
